test/pthread/test_pthread_memory_growth_mainthread.c

```diff
@@ -4,51 +4,101 @@
 // found in the LICENSE file.
 
 #include <pthread.h>
-#include <sys/types.h>
-#include <stdio.h>
 #include <stdlib.h>
 #include <assert.h>
-#include <emscripten/em_asm.h>
+#include <emscripten/em_js.h>
+#include <stdbool.h>
 
-static void *thread_start(void *arg)
+// We want to test that JavaScript access in a pthread automatically updates the
+// memory views if the heap has grown on the main thread meanwhile.
+//
+// Checking this correctly is somewhat tricky because a lot of Emscripten APIs
+// access heap on their own, so the memory views might be updated by them before
+// we explicitly check the state in our own JS routines below, which leads to
+// false positives (test passing even though in isolation the heap access in our
+// own JS is not the one updating the memory views).
+//
+// To test it in isolation, we need to use only EM_JS - which is as close to a
+// pure JS call as we can get - and not EM_ASM, pthread_join, proxying etc., not
+// even `puts`/`printf` for logging - as all of those access the heap from JS on
+// their own and might update the memory views too early. Not using standard
+// proxying mechanisms also means we need to drop down all the way to raw
+// atomics.
+
+EM_JS(void, js_assert_initial_heap_state, (bool isWorker), {
+  console.log(`Checking initial heap state on the ${isWorker ? 'worker' : 'main'} thread`);
+  assert(HEAP8.length === 32 * 1024 * 1024, "start at 32MB");
+});
+
+EM_JS(void, js_assert_final_heap_state, (bool isWorker, const char* buffer, int finalHeapSize), {
+  console.log(`Checking final heap state on the ${isWorker ? 'worker' : 'main'} thread`);
+  assert(HEAP8.length > finalHeapSize, "end with >64MB");
+  assert(HEAP8[buffer] === 42, "readable from JS");
```

Comment:
Perhaps stash a reference to HEAP8 initially, then compare it later and see it changed?

Reply:
Could do, sure.

Although the length checks already ensure growth and should be somewhat more reliable if/when the resizable ArrayBuffer comes to Wasm, as in that case the heap reference will remain the same but the size should still change.

Reply:
I think I'm going to keep it as-is for reasons above. Stashing a reference doesn't add new value but does make test potentially less reliable in the future.